contracts/basic-bandwidth-generation/contracts/BandwidthGenerator.sol

```diff
@@ -0,0 +1,91 @@
+pragma solidity 0.6.6;
+
+import "./CosmosToken.sol";
+import "./Gravity.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+/** 
+ * @title BandwidthGenerator
+ * @dev   Contract for generating Basic Bandwidth Credentials (BBCs) on the Nym cosmos blockchain, 
+ *        using ERC20 representations of NYM as payment. Utilises the Gravity Bridge for cross-chain payment. 
+ *        
+ *        Credentials represent a certain amount of bandwidth which can be sent through the Nym Mixnet. 
+ *        By default 1 NYM = 1 GB of bandwidth. The `MBPerToken` amount can be adjusted by the contract owner. 
+ * 
+ *        The amount of bandwidth bought is calculated according to the following formula: 
+ *        `(Token amount in 'wei' / 10**18) * MBPerToken`
+ */ 
+contract BandwidthGenerator is Ownable {
+    
+    CosmosERC20 public erc20;
+    Gravity     public gravityBridge; 
+    uint256     public MBPerToken; 
+    
+    event BBCredentialPurchased(
+        uint256 indexed Bandwidth,
+        uint256 indexed VerificationKey,
+        bytes   SignedVerificationKey, 
+        bytes32 indexed CosmosRecipient
+    );
+
+    event RatioChanged(
+        uint256 indexed NewMBPerToken
+    );
+    
+    /**
+     * @param _erc20          Address of the erc20NYM deployed through the Gravity Bridge.
+     * @param _gravityBridge  Address of the deployed Gravity Bridge. 
+     */
+    constructor(CosmosERC20 _erc20, Gravity _gravityBridge) public {
+        require(address(_erc20) != address(0),         "BandwidthGenerator: erc20 address cannot be null"); 
+        require(address(_gravityBridge) != address(0), "BandwidthGenerator: gravity bridge address cannot be null"); 
+        erc20 = _erc20;
+        gravityBridge = _gravityBridge; 
+        MBPerToken = 1024; // default amount set at deployment: 1 erc20NYM = 1024MB = 1GB
```

Comment:
This is making the assumption that we operate on MB level, whereas the client and gateway currently count bandwidth at byte level. I think this should be `BytesPerToken = 1024 * 1024 * 1024` so that the contract counts at byte level as well. That would be better than working with floats in the client/gateway, as the Sphinx packets are smaller than MB level from what I recall.